### ISSUE TITLE
Export to MIDI

### ISSUE THREAD

**Issue by daveyarwood** [alda-lang/alda-core#9]
Moved from https://github.com/alda-lang/alda/issues/145.

Quoting myself: 

> MIDI export should be a separate mode of use for Alda, i.e. rather than _playing_ a score (which will use Alda's event scheduling system to play the score in real time), you can _export_ it to MIDI.
>
> I like the idea of reusing the machinery we have in place as much as we can -- the score would still parse the same way into [alda.lisp](https://github.com/alda-lang/alda/blob/master/doc/alda-lisp.md) code, and we would still evaluate it in order to produce a map of score data. But from there, we'll implement a totally new system that will take the score map and use it to produce a MIDI file. If we need to, we can add additional context to the score map via the functions in alda.lisp (e.g. adding fields to the note event). But, I think it should be fairly doable to take the information we already have in the score map and use it to generate a MIDI file.

**Comment by pierredewilde** [comment]
Exporting to MIDI will be very cool. 

Alda scores (converted to MIDI) could be used with any MIDI sound modules or synthesizers. Moreover, Alda scores could be converted to sheet music with MuseScore or related music notation software.

**Comment by kfrncs** [comment]
Do you have an ETA on this feature? looking forward to trying it out. Would love to use Alda with my Microkorg.

**Comment by daveyarwood** [comment]
No ETA in particular, but I am hopeful that we can have at least a "rough cut" of this feature in the near future. @jgkamat and @oakes have done some excellent work over at https://github.com/alda-lang/alda-sound-engine-clj/pull/11 that will enable us to export MIDI in a convenient way. There are just a handful of small issues to hammer out ([checklist](https://github.com/alda-lang/alda-sound-engine-clj/pull/11#issuecomment-379139795)). Contributions would be welcome, if anyone feels like jumping in to help get this over the finish line!

Once we can merge that PR, it will mean that the internal implementation of Alda's sound generator will be based on the Java MIDI Sequencer. And at that point, it will be trivial to export MIDI.

**Comment by 0atman** [comment]
Could anything be learned from https://github.com/oakes/edna/blob/master/src/edna/core.clj#L229 (or be taught directly by @oakes ?

**Comment by oakes** [comment]
I'm still having some trouble getting edna's export working on sound-engine-clj 1.0.0. I'm guessing it's the parameters I'm passing to the AudioFormat constructor but I haven't had much time lately to figure out what needs to change. Once that is fixed it should be easy to transfer the code into a pure alda project, and all you'd need to add is the meico dependency.

**Comment by daveyarwood** [comment]
I think we're very close to having support for exporting to MIDI in Alda, with the latest changes in https://github.com/alda-lang/alda-sound-engine-clj

@TBuc and I have been chatting about it a bit in the #development channel in [Slack](https://slack.alda.io). It will take a little bit of work, but not much.

Here's a little excerpt of my off-the-cuff scribbled notes from that Slack conversation:

>@TBuc we have this (currently unused) function `export-midi-file!` in alda-sound-engine-clj: https://github.com/alda-lang/alda-sound-engine-clj/blob/eb52896/src/alda/sound/midi.clj#L297-L299
>
>it takes a javax.sound.midi.Sequence and the desired filename of the saved midi file
>
>Sequence is something that we use internally in the sound engine now. so, i think the next step would be to implement a worker command in alda-server-clj here: https://github.com/alda-lang/alda-server-clj/blob/master/src/alda/worker.clj
>
>the implementation would basically be calling that function when the worker gets an `"export"` command with the appropriate options (i.e. export type is midi, filename is some non-empty string) and then returning a success response
>
>after that, it would just be a matter of hooking it up in the client, which i have some ideas about. tl;dr the syntax would look something like `alda export -f somefile.alda -t midi -o somefile.mid`
>
>granted, it's a bit more complicated than just calling that function
>we'll have to parse the score and create the sequence
>but the implementation is already there for those things
>just a matter of hooking everything up
>i'm being hand-wavey because i'm a little busy at the moment, but definitely ping me if you get stuck!

**Comment by daveyarwood** [comment]
A thought I just had: maybe after we have support for exporting to MIDI, a better approach for "export to WAV" would be to find an existing CLI tool that will export MIDI -> WAV and let that do the work. We could include a good recipe in our docs showing how to do it, but we wouldn't necessarily need to worry about handling all of the edge cases or including another library as a dependency in order to support WAV export natively in Alda.

**Comment by pierredewilde** [comment]
Exporting Alda to MIDI is great but it would be even greater to play Alda on MIDI devices in realtime by specifying the name of MIDI device to use.

**Comment by 0atman** [comment]
This should not be too difficult for someone familiar with java midi - it's simply a matter of redirecting the midi output to a different device - might even be a 1-liner!. @pierredewilde would you like to give it a go? :laughing:

**Comment by pierredewilde** [comment]
Sorry, Java is not my cup of tea :)

**Comment by 0atman** [comment]
Good answer! :coffee:

**Comment by daveyarwood** [comment]
Summarizing some thoughts I had after tinkering with MIDI export for a while:

* We are currently working with MIDI sequences where the division type is SMPTE.
  * In layman's terms, SMPTE works in terms of absolute time and does not have a "tempo."
  * By contrast, PPQ (pulses per quarter note) expresses notes in relation to the quarter note, and you can set the tempo.
  * There is no way to set the tempo of a SMPTE MIDI sequence. The ticks are related to frames per second.

* Under the hood, Alda does not actually work in terms of BPM, but rather lengths of time in milliseconds. This allows Alda composers to express note lengths in terms of seconds/milliseconds, as well as standard note lengths (which Alda then translates into milliseconds).

* I was able to take an alda score and export it to a `.mid` file, but then when i played it using timidity, it played super fast because, _I think_, timidity doesn't properly support SMPTE MIDI files.

* I imported the midi file into MuseScore and got absolute garbage that sounded nothing like the score. The notes were all jumbled up and not the right speed at all.

* I started looking for MIDI analysis tools and found [MidiYodi](http://www.canato.se/midiyodi/). Interestingly enough, I opened the MIDI file in MidiYodi and played it, and the timing sounded exactly correct.

* So, I think the issue is that a lot of software out there that deals with MIDI doesn't properly handle SMPTE. That sort of makes sense to me because I think in a lot of cases, people write music with standard note lengths in mind, not lengths of notes in milliseconds.

* The lack of support in other software for SMPTE MIDI is unfortunate because it probably means that it won't be feasible to take a score in Alda and import it into, say, Sibelius or Finale.

* I think we might want to consider working with scores in BPM and therefore use PPQ (pulses per quarter note) instead of SMPTE. We could probably convert the non-standard notes (i.e. notes expressed in seconds/milliseconds) in an Alda score into the right number of ticks, such that most software that works with MIDI could figure out how to represent the note in standard musical notation.

* Maybe the best thing would be for Alda to support both modes -- SMPTE vs. PPQ -- and allow the user to choose which one makes sense for what he/she is trying to do with the output MIDI file?

**Comment by elydpg** [comment]
Probably. Out of curiosity, are notes in alda precise only to the millisecond?

**Comment by daveyarwood** [comment]
They are more precise than that, as the numbers of milliseconds can be floating-point.

That being said, I think Alda 2.0 will likely end up being precise only to the millisecond, in order to simplify the communication between client and player. I think this should be totally fine, as the human ear isn't capable of recognizing differences of several milliseconds, let alone less than 1 millisecond.

**Comment by daveyarwood** [comment]
More shower thoughts:

* One thing that Alda lets you do that goes against the grain of MIDI is that you can have multiple instruments all playing at different tempos. We are able to do this because Alda measures timing in milliseconds, not beats.

* This might be an argument for SMPTE by default, however I think PPQ should probably be the default because that's what the user will expect 99% of the time. In most cases, a piece of music has a tempo that is global and every instrument part is playing at the same tempo.
  * In most cases, producing a PPQ MIDI file is more useful because it allows you to import that MIDI file into e.g. sheet music software like MuseScore, and the sheet music software will be aware of the tempo of the score and therefore be able to accurately group notes into measures.
    * (If Alda provided a way to specify time signatures, that would give sheet music software better context to group notes into measures correctly. That's something we should also do in the future.)

* Some changes to the score updating machinery will be required in order to support this:
  * We can continue to keep track of each instrument's offset in milliseconds, but we'll also keep track of the instrument's offset in _beats_.
  * Each note event will include its offset and duration in both milliseconds and beats, as well.
  * We will keep track of the "master" / "global" tempo and when (in absolute terms, i.e. milliseconds) it changes during the score. This will allow us to specify the tempo in the MIDI sequence, which is required in order to support PPQ.

* I think we can still support exporting to MIDI in cases when the score has instruments playing in different tempos at the same time. We just have to pick one (let's say the first declared) instrument and have that be the source of truth for the "true" tempo, i.e. the tempo that we will set on the MIDI sequence. Then, if other instruments are playing in different tempos, we can record their events' offsets in beats by adjusting them in proportion to the "master" tempo.
  * For example, if instrument A is playing at 100 bpm and instrument B is playing a little faster at 120 bpm, then I think we could use the ratio 100:120 to adjust instrument B's notes' lengths (in beats) to be shorter in the master tempo of 100 bpm.
  * Note that the events' offsets in milliseconds will remain exactly the same, since they are absolute and not tied to any particular tempo.

* Once we have these changes to the score updating machinery in place, we can update the sound engine code that creates the MIDI sequence:
  * use PPQ instead of SMPTE
  * include tempo change events in the sequence
  * open question: is there a way to express the "starting" tempo? or do you just put a tempo change event at the beginning of the sequence?

**Comment by pierredewilde** [comment]
`(tempo! 120)` at the start of sequence should be enough.

Time expressed in PPQ and/or SMPTE is only relevant for exporting MIDI files. 

If you plan to send MIDI commands directly to external MIDI devices (in realtime), you need to wait (in ms) between MIDI commands.

Eg
```
[ 144, 60, 100 ]
wait 100ms 
[ 128, 60, 0 ]
```

**Comment by daveyarwood** [comment]
This feature is now available in [Alda 1.3.0](https://github.com/alda-lang/alda/releases/tag/1.3.0)! :tada:

**Comment by 0atman** [comment]
AMAZING!
![tenor](https://user-images.githubusercontent.com/114097/56133774-2f474600-5f85-11e9-9021-f852531074fe.gif)
